### PR TITLE
Additional database file checks in cli/Utils.unlockDatabase

### PR DESCRIPTION
**Pull request by sjamesr**
[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
- ✅ New feature (non-breaking change which adds functionality)

## Description and Context
Avoids prompting the user for a password if unlocking is likely to fail
due to some problem with the database file (i.e. not found, not a file,
not readable).

## Testing strategy
Ran tests, tested manually.

## Checklist:
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

**Comment by sjamesr** [comment]
This is separate from #3426 because the old behavior already exists in `keepassxc-cli`.

File: src/cli/Utils.cpp [hunk 1]
```diff
@@ -108,6 +109,22 @@ namespace Utils
         TextStream out(outputDescriptor);
         TextStream err(errorDescriptor);
 
+        QFileInfo dbFileInfo(databaseFilename);
```
Comment (droidmonkey) [inline src/cli/Utils.cpp]: Might want to throw in here a check of the canonical file path. That function call returns empty string if the fully resolved file does not exist.

Comment (sjamesr) [reply src/cli/Utils.cpp]: Interestingly, with the following patch:

```patch
diff --git a/src/cli/Utils.cpp b/src/cli/Utils.cpp
index b686b155..fede0b29 100644
--- a/src/cli/Utils.cpp
+++ b/src/cli/Utils.cpp
@@ -110,7 +110,9 @@ namespace Utils
         TextStream err(errorDescriptor);
 
         QFileInfo dbFileInfo(databaseFilename);
-        if (!dbFileInfo.exists()) {
+        qInfo("Path: %s", databaseFilename.toStdString().c_str());
+        qInfo("Canonical path: %s", dbFileInfo.canonicalPath().toStdString().c_str());
+        if (dbFileInfo.canonicalPath().isEmpty()) {
             out << QObject::tr("Failed to open database file %1: not found").arg(databaseFilename) << endl;
             return {};
         }
```

I get

```
$ ./src/cli/keepassxc-cli show /tmp/bar foo
Path: /tmp/bar
Canonical path: .
Failed to open database file /tmp/bar: not a plain file
```

So `canonicalPath` returns `.` for a file that does not exist? Doesn't agree with the documentation.

Comment (droidmonkey) [reply src/cli/Utils.cpp]: You want this function: https://doc.qt.io/qt-5/qfileinfo.html#canonicalFilePath

Comment (sjamesr) [reply src/cli/Utils.cpp]: That works, thanks.

File: src/cli/Utils.cpp [hunk 2]
```diff
@@ -108,6 +109,22 @@ namespace Utils
...
+        if (dbFileInfo.canonicalFilePath().isEmpty()) {
+            out << QObject::tr("Failed to open database file %1: not found").arg(databaseFilename) << endl;
+            return {};
+        }
```
Comment (louib) [inline src/cli/Utils.cpp]: @sjamesr could we add unit tests on some of those new checks?

Comment (sjamesr) [reply src/cli/Utils.cpp]: Done, there should be coverage for all the checks now.

Comment (louib) [reply src/cli/Utils.cpp]: Tests are failing on Windows now
```
FAIL!  : TestCli::testInvalidDbFiles() Compared values are not the same
    Actual   (QString(m_stderrFile->readAll()))                                             : "Error while reading the database: Failed to read database file.\n"
    Expected (QObject::tr("Failed to open database file %1: not readable").arg(path) + "\n"): "Failed to open database file C:/msys64/tmp/testcli.FULFwe: not readable\n"
```

Comment (sjamesr) [reply src/cli/Utils.cpp]: I couldn't get isReadable to return false on Windows, even with NTFS permissions enabled. I left the check in but disabled the test on Windows.

**Pull request merged**